0024-pre-shared-key-v2/0024-pre-shared-key-v2.md

```diff
@@ -0,0 +1,266 @@
+7. Receiver creates a [Payment Chunk Response](#payment-chunk-response) with the same Payment ID and Sequence number. The Payment Amount and Chunk Amount are set to the amount that arrived on the incoming transfer. The receiver then [encrypts](#encryption-envelope) the data.
+8. Receiver generates fulfillment as described [below](#fulfillment-generation) and fulfills the incoming transfer with the fulfillment and the encrypted response data.
+9. Sender is notified of the outgoing transfer being fulfilled and can use the decrypted response to determine the amount that arrived at the destination.
+
+### Chunked Payment
+
+1. Sender creates a packet with Type 0, a random Payment ID, a Sequence number of 0 and a Chunk Amount of 0. In the case of a fixed destination amount, sender sets the Payment Amount to the destination amount. In the case of a fixed source amount, sender sets the Payment Amount to 18446744073709551615 (the maximum possible value). Sender [encrypts the request](#encryption-envelope).
+2. Sender generates the payment condition as described [below](#fulfillable-condition).
+3. Sender sends a payment with the condition from step 2., the encrypted request from step 1. as the data, and an arbitrary source amount.
+4. Receiver is notified of the incoming transfer and decrypts the data.
+5. Receiver checks that the incoming transfer amount is greater than or equal to the Chunk Amount and that the total that has arrived thus far does not exceed the Payment Amount.
+6. Receiver creates a response packet with the Type 2, the same Payment ID, Sequence. The Payment Amount is set to the total the receiver has received and fulfilled thus far (including the current chunk). Receiver sets the Chunk Amount to the amount that arrived on the incoming transfer. Receiver [encrypts the response](#encryption-envelope).
+7. Receiver generates fulfillment as described [below](#fulfillment-generation) and fulfills the incoming transfer.
+8. Sender is notified of the outgoing transfer being fulfilled and decrypts the attached data.
+9. Sender determines the exchange rate by dividing the Chunk Amount from the receiver by the chunk's source amount.
+10. Sender repeats step 1-8 until the fixed source or destination amount has been sent or delivered, respectively. Sender MAY adjust the chunk amount upwards or downwards in response to successful and unsuccessful payments. The sender SHOULD set the Type to 1 on the last chunk.
+
+## Specification
+
+### Encryption
+
+Request and response data are encrypted using [AES-256-GCM](https://en.wikipedia.org/wiki/Galois/Counter_Mode) with a 16-byte Initialization Vector (IV).
+
+#### Encryption Envelope
+
+```
+(Numbers represent bytes)
+
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
```

Comment:
If numbers represent bytes, this should start counting at 1. Also, it should not start over at 0 but should keep going 10, 11, etc.

Maybe better to use powers of 2 instead.